### DISCUSSION TITLE
XFS structure needs cleaning, Moose error opening file.

### DISCUSSION THREAD

**Discussion by bighippo999** [moosefs/moosefs#345] [Q&A]
## Have you read through available documentation and open Github issues?
Yes

## Is this a BUG report, FEATURE request, or a QUESTION? Who is the indended audience?
Bug, maybe.

## System information
7x Raspbberry Pi 4(4gb), 1 as metaserver and 6 with 5TB drives, (OS 64GB Partition EXT4, Data 4.5TB Partition XFS). Latest Raspbian OS (Raspbian GNU/Linux 10 (buster)) and all software updates.

#### Your moosefs version and its origin (moosefs.com, packaged by distro, built from source, ...).
3.0.111, following guide from https://moosefs.com/download/#current, using stretch source. I think the initial version was 3.0.109 and upgraded with apt-get upgrade

#### Operating system (distribution) and kernel version.
Distributor ID: Raspbian
Description:    Raspbian GNU/Linux 10 (buster)
Release:        10
Codename:       buster

#### Hardware / network configuration, and underlying filesystems on master, chunkservers, and clients.
Hardware above. Network Gb switch.

#### How much data is tracked by moosefs master (order of magnitude)?
 - All fs objects: 906485
 - Total space: 27 TiB
 - Free space: 12 TiB
 - RAM used: 413 MiB
 - last metadata save duration: ~1.9Secs

## Describe the problem you observed.
Error reported on cgi page against chunkserver #4, but no indication what was wrong. After digging around and using the script (slightly modified) from https://github.com/moosefs/moosefs/issues/106 I was able to see 2 chunks with problems:
/mnt/mfschunks1/00/chunk_00000000000BF267_00000001.mfs
/mnt/mfschunks1/07/chunk_0000000000061BB5_00000001.mfs

Running the mfschunktool against these 2 files reported: 'error opening file'.
ls -lh against these 2 files gives: Structure needs cleaning.

So an underlying filesystem problem.

stopped moose-master and all chunkservers. unmount /mnt/mfschunks1 from #4 and ran xfs_repair
against the disk. 'bad CRC for inode 46001118' and it repaired the errors.

It wiped out those 2 chunks when repairing though.

Remount it and I removed the .chunksdb from #4 to force it to scan on start, which it did. Then start all the others and the master. cgi showed 2 chunks undergoal/endangered (not sure which can't remember. goal 2: Valid 1).
Moose then did exactly what it was supposed to and replicate those out.

The problem is: The error had been there for a few days and those 2 chunks were technically undergoal/endangered all those days having only 1 copy that was accessible due to errors on the other copy, which it knew something of.

So I'm not sure if you'd class it as a bug. If not then a suggestion that when there's an error on a chunk it still aims to keep the goal. I think being overgoal is safer the under :)

I'm pretty sure that the fault was caused by a power loss (my fault). I have another problem related to that power issue: 2 chunks (goal 0, valid 0) ready to be removed. Which I assume are the ones showing as locked-unused and different to the above 2 as it's still showing. I'm hoping that just fixes itself after a month (2 weeks so far), I'm sure I've read that rogue chunks from other servers are kept for 30 days. While they're not from another moose instance, I think the power failure may have happened as data was writing and the master just doesn't know what they are. And yes, the whole setup is now on UPS with nut configured to safely shutdown.

#### Can you reproduce it? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.
No idea how I could force xfs to screw up on purpose.

#### Include any warning/errors/backtraces from the system logs
Can't see anything in the logs:
grep -i BF267 /var/log/moose.log (or the other chunk 61BB5) doesn't return anything on either the master or the affected chunk server.

**Comment by borkd** [comment]
If you operate on IoT class hardware rather than enterprise servers, and your data is not disposable, I would recommend a conservative stance:
- turn off write cache on any disk used with MFS
- tell MFS to use fsync
- use XFS with `crc=1`
- rid of .chunkdb before chunkserver starts
- Decide how long a full sweep should take matching the I/O capability of the each chunkserver to the number of chunks it is expected to carry. Skew the configuration for chunk checking accordingly.

Regarding writing a test that "screws xfs on purpose" there are some options. Typical approach:
- configure lossy block device(s) on as many machines as needed
  - scsi_debug works well
  - dmsetup also can create a finicky block device if you use the "error" target
- format the devices with the filesystem of choice
- launch distributed filesystem components you want to examine
- start writing data, or simulating failures elsewhere that result in replication kicking in
- grab some popcorn and wait for the meltdown
- file out new bug reports

If you want to take this misery any further, you can employ chaos engineering tools between instances to simulate partial or complete network outages, time warps, and other nasties. 

Experience may be eye opening and a bit traumatic, so it should go without saying all this should be done on a disposable cluster and not anywhere near your live data.

**Comment by chogata** [comment]
@bighippo999 

MooseFS chunk server does scan all the chunks regularly and if it finds a chunk with problems (for example, with read errors), it _immediately_ reports the problem to master server and such a chunk is replicated. 
But of course the cycle of scanning all chunks takes time, because it can't be done too fast, otherwise it would take too many resources and chunk server would not be able to perform its regular operations. The bigger the number of chunks, the more time passes between consecutive scans of the same chunk.
In your case the most probable scenario is that the chunk server found a chunk with a problem, reported the problem to master, and the master reacted (i.e. ordered replication) and displayed an error message for this chunk server. You saw the message and you started looking by yourself and you found two more chunks with problems, before the chunk server had time to do so. Chunk server would have found them and would have replicated them eventually. But you were faster ;)

It's good that you scanned your underlying XFS, after a power failure on a chunk server it's always a good idea to scan all local filesystems on storage disks before running the chunk server again.

**Comment by bighippo999** [comment]
@borkd Thanks for the info, dammit now I have a reason to buy even more PI's for testing :) The data isn't exactly disposable, but it is replicate elsewhere so can be recovered but I'm looking into implementing your suggestions this weekend (free time yay).

@chogata That is a reasonable scenario and I've no way of knowing if that's what happened then. 
but sadly #4 has report an error again today :( '2020-03-05 08:12:47 on chunk: 11678' looking at the logs it also found another problem 2 days ago (I don't look at the gui daily).
2 chunks, same problem as before (can't open file), but it's given me the chance to look into it a bit more and using find (find /mnt/mfschunks1/ -name 'chunk_0000000000002D9E_00000001.mfs') on all the other servers I can see #3 and #6 have copies of the one chunk and #1 and #5 on the other chunk. So it looks like you were correct :)

I just have to figure out why #4 seems to be the problem server. Time to go looking for badblocks I guess :(

As I said they are now behind an UPS with NUT configured, so hopefully no more power issues at least.
